src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs

```diff
@@ -223,38 +223,22 @@ public static void ThrowJsonException_JsonRequiredPropertyMissing(JsonTypeInfo p
 
             Debug.Assert(parent.PropertyCache != null);
 
-            int propertyIdx = 0;
-
-            for (int requiredPropertyIdx = 0; requiredPropertyIdx < missingJsonProperties.Length; requiredPropertyIdx++)
+            for (int propertyIdx = 0; propertyIdx < parent.PropertyCache.List.Count; propertyIdx++)
             {
-                if (!missingJsonProperties[requiredPropertyIdx])
-                {
-                    continue;
-                }
+                JsonPropertyInfo property = parent.PropertyCache.List[propertyIdx].Value;
 
-                JsonPropertyInfo? missingProperty = null;
-
-                // requiredPropertyIdx indices occur consecutively so we can resume iteration
-                for (; propertyIdx < parent.PropertyCache.List.Count; propertyIdx++)
+                if (!property.IsRequired || !missingJsonProperties[property.RequiredPropertyIndex])
                 {
-                    JsonPropertyInfo maybeMissingProperty = parent.PropertyCache.List[propertyIdx].Value;
-                    if (maybeMissingProperty.IsRequired && maybeMissingProperty.RequiredPropertyIndex == requiredPropertyIdx)
-                    {
-                        missingProperty = maybeMissingProperty;
-                        break;
-                    }
+                    continue;
                 }
 
-                Debug.Assert(propertyIdx != parent.PropertyCache.List.Count);
-                Debug.Assert(missingProperty != null);
-
                 if (!first)
                 {
                     listOfMissingPropertiesBuilder.Append(CultureInfo.CurrentUICulture.TextInfo.ListSeparator);
                     listOfMissingPropertiesBuilder.Append(' ');
                 }
 
-                listOfMissingPropertiesBuilder.Append(missingProperty.Name);
+                listOfMissingPropertiesBuilder.Append(property.Name);
```

Comment:
Would it make sense to introduce a cutoff to the number of reported missing properties? I think the message becomes counterproductive as soon as the missing properties exceed 3 or 4 entries.

Reply:
possibly soft length check? i.e. we finish once we exceed 50/100 characters on the string builder. Some properties are short so it might make sense to show more of them. I personally don't mind long messages so I'm ok with leaving this as is

Reply:
That could work. Just need to make sure that the exception message is readable under all circumstances and the stacktrace is not obfuscated. You might need to update the error message to indicate that the list is not exhaustive.

Reply:
truncating to 50 characters, that should give room for couple of properties without being excessively long